examples/websockets/src/main.rs

```diff
@@ -59,58 +76,169 @@ async fn main() {
     let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
     tracing::debug!("listening on {}", addr);
     axum::Server::bind(&addr)
-        .serve(app.into_make_service())
+        .serve(app.into_make_service_with_connect_info::<SocketAddr>())
         .await
         .unwrap();
 }
 
+/// A formally required handler for the request.
+/// At this point we can extract useful TCP/IP metadata such as IP address of the client
+/// as well as user-agent of the browser
 async fn ws_handler(
     ws: WebSocketUpgrade,
     user_agent: Option<TypedHeader<headers::UserAgent>>,
+    ConnectInfo(addr): ConnectInfo<SocketAddr>,
 ) -> impl IntoResponse {
-    if let Some(TypedHeader(user_agent)) = user_agent {
-        println!("`{}` connected", user_agent.as_str());
-    }
-
-    ws.on_upgrade(handle_socket)
+    let user_agent = if let Some(TypedHeader(user_agent)) = user_agent {
+        user_agent.to_string()
+    } else {
+        String::from("Unknown browser")
+    };
+    println!("`{}` at {} connected.", user_agent, addr.to_string());
+    // finalize the upgrade process by returning upgrade callback.
+    // we can customize the callback by sending additional info such as address.
+    ws.on_upgrade(move |socket| handle_socket(socket, addr))
 }
 
-async fn handle_socket(mut socket: WebSocket) {
+/// Actual websocket statemachine (one will be spawned per connection)
+async fn handle_socket(mut socket: WebSocket, who: SocketAddr) {
+    //send a ping (unsupported by some browsers) just to kick things off and get a response
+    if let Ok(_) = socket.send(Message::Ping(vec![1, 2, 3])).await {
+        println!("Pinged {}...", who);
+    } else {
+        println!("Could not ping {}!", who);
+        return;
+    }
+
+    // receive single message form a client (we can either receive or send with socket)
+    // this will likely be the Pong for our Ping or a hello.
     if let Some(msg) = socket.recv().await {
         if let Ok(msg) = msg {
-            match msg {
-                Message::Text(t) => {
-                    println!("client sent str: {:?}", t);
-                }
-                Message::Binary(_) => {
-                    println!("client sent binary data");
-                }
-                Message::Ping(_) => {
-                    println!("socket ping");
-                }
-                Message::Pong(_) => {
-                    println!("socket pong");
-                }
-                Message::Close(_) => {
-                    println!("client disconnected");
-                    return;
-                }
+            if process_message(msg, who) {
+                return;
             }
         } else {
-            println!("client disconnected");
+            println!("{} abruptly disconnected", who);
```

Comment:
Why "abruptly"? Isn't `client disconnected` or `{who} disconnected` fine?

Reply:
I think that in order to get into that branch the client would need to have disconnected in some *very strange way*. So by no means it would be a "normal business as usual event". However, in all my attempts I was unable to ever get into that error branch, no matter how I tortured the client or network (which included pulling network interface and killing client), so I do not even know how one would get that to occur.

Reply:
Alright I see. We can just keep "abruptly" I guess then.